File: libraries/RPGTableHelper.BusinessLayer/QueryHandlers/JsonDiffer.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace RPGTableHelper.BusinessLayer.QueryHandlers;

public static class JsonDiffer
{
    public static List<JsonObject> Diff(JsonNode oldJson, JsonNode newJson)
    {
        var patches = new List<JsonObject>();

        int oldVersion = oldJson?["version"]?.GetValue<int>() ?? throw new Exception("Old JSON must have version");
        int newVersion = newJson?["version"]?.GetValue<int>() ?? throw new Exception("New JSON must have version");

        if (oldVersion == newVersion)
        {
            throw new Exception("Versions must differ for a valid patch");
        }

        // Add version update
        patches.Add(
            new JsonObject
            {
                ["op"] = "version",
                ["from"] = oldVersion,
                ["to"] = newVersion,
            }
        );

        GenerateDiff(oldJson, newJson, string.Empty, patches);

        // remove patches for version
        patches.RemoveAll(p => p["op"]?.ToString() == "replace" && p["path"]?.ToString() == "/version");

        return patches;
    }

    public static JsonNode Apply(JsonNode original, List<JsonObject> operations)
    {
        if (operations == null || operations.Count == 0)
        {
            throw new Exception("Empty patch");
        }

        var versionOp = operations[0];
        if (versionOp["op"]?.ToString() != "version")
        {
            throw new Exception("First operation must be a version update");
        }

        int expectedVersion = versionOp["from"]?.GetValue<int>() ?? throw new Exception("Missing version 'from'");
        int targetVersion = versionOp["to"]?.GetValue<int>() ?? throw new Exception("Missing version 'to'");

        if (original?["version"]?.GetValue<int>() != expectedVersion)
        {
            throw new Exception($"Version mismatch. Expected {expectedVersion}");
        }

        var result = original?.DeepClone() ?? new JsonObject();

        foreach (var op in operations.Skip(1))
        {
            string opType = op["op"]?.ToString() ?? throw new Exception("Missing op type");
            string path = op["path"]?.ToString() ?? throw new Exception("Missing path");

            ApplyOperation(result, opType, path, op);
        }

        result["version"] = targetVersion;
        return result!;
    }

    private static void ApplyOperation(JsonNode? root, string op, string path, JsonNode opData)
    {
        var parts = path.Trim('/').Split('/');
        var current = root;

        for (int i = 0; i < parts.Length - 1; i++)
        {
            if (current is JsonObject obj)
            {
                current = obj[parts[i]];
            }
            else if (current is JsonArray arr && int.TryParse(parts[i], out int index) && index < arr.Count)
            {
                current = arr[index];
            }
            else
            {
                throw new Exception($"Invalid path: {path}");
            }
        }

        var lastPart = parts[^1];

        if (current is JsonObject parentObj)
        {
            switch (op)
            {
                case "add":
                case "replace":
                    parentObj[lastPart] = opData["value"]!.DeepClone();
                    break;
                case "remove":
                    parentObj.Remove(lastPart);
                    break;
                default:
                    throw new Exception($"Unsupported op: {op}");
            }
        }
        else if (current is JsonArray parentArr)
        {
            if (!int.TryParse(lastPart, out int index))
            {
                throw new Exception($"Invalid array index: {lastPart}");
            }

            switch (op)
            {
                case "add":
                    if (index > parentArr.Count)
                    {
                        throw new Exception($"Invalid add index: {index}");
                    }

                    parentArr.Insert(index, opData["value"]!.DeepClone());
                    break;
                case "replace":
                    if (index >= parentArr.Count)
                    {
                        throw new Exception($"Invalid replace index: {index}");
                    }

                    parentArr[index] = opData["value"]!.DeepClone();
                    break;
                case "remove":
                    if (index >= parentArr.Count)
                    {
                        throw new Exception($"Invalid remove index: {index}");
                    }

                    parentArr.RemoveAt(index);
                    break;
                default:
                    throw new Exception($"Unsupported op: {op}");
            }
        }
        else
        {
            throw new Exception($"Unsupported parent type for path: {path}");
        }
    }

    private static void GenerateDiff(JsonNode? oldNode, JsonNode? newNode, string path, List<JsonObject> patches)
    {
        if (oldNode is JsonObject oldObj && newNode is JsonObject newObj)
        {
            foreach (var prop in oldObj)
            {
                if (!newObj.ContainsKey(prop.Key))
                {
                    patches.Add(new JsonObject { ["op"] = "remove", ["path"] = $"{path}/{prop.Key}" });
                }
                else
                {
                    GenerateDiff(prop.Value, newObj[prop.Key], $"{path}/{prop.Key}", patches);
                }
            }

            foreach (var prop in newObj)
            {
                if (!oldObj.ContainsKey(prop.Key))
                {
                    patches.Add(
                        new JsonObject
                        {
                            ["op"] = "add",
                            ["path"] = $"{path}/{prop.Key}",
                            ["value"] = prop.Value!.DeepClone(),
                        }
                    );
                }
            }
        }
        else if (oldNode is JsonArray oldArr && newNode is JsonArray newArr)
        {
            int commonLength = Math.Min(oldArr.Count, newArr.Count);
            for (int i = 0; i < commonLength; i++)
            {
                GenerateDiff(oldArr[i], newArr[i], $"{path}/{i}", patches);
            }

            for (int i = commonLength; i < newArr.Count; i++)
            {
                patches.Add(
                    new JsonObject
                    {
                        ["op"] = "add",
                        ["path"] = $"{path}/{i}",
                        ["value"] = newArr[i]!.DeepClone(),
                    }
                );
            }

            for (int i = oldArr.Count - 1; i >= newArr.Count; i--)
            {
                patches.Add(new JsonObject { ["op"] = "remove", ["path"] = $"{path}/{i}" });
            }
        }
        else if (!JsonNode.DeepEquals(oldNode, newNode))
        {
            patches.Add(
                new JsonObject
                {
                    ["op"] = "replace",
                    ["path"] = path,
                    ["value"] = newNode!.DeepClone(),
                }
            );
        }
    }
}

File: applications/rpg_table_helper/lib/helpers/json_differ.dart
import 'dart:convert';

class JsonDiffer {
  /// Diffs two versioned JSON blobs and returns a patch.
  /// Optionally includes current and new version numbers.
  static List<Map<String, dynamic>> diff(
    Map<String, dynamic> oldJson,
    Map<String, dynamic> newJson,
  ) {
    final oldVersion = oldJson['version'] as int?;
    final newVersion = newJson['version'] as int?;

    if (oldVersion == null || newVersion == null) {
      throw Exception("Missing 'version' key in JSON");
    }

    final patch = _diff(oldJson, newJson);

    patch.insert(0, {
      'op': 'version',
      'from': oldVersion,
      'to': newVersion,
    });

    return patch;
  }

  /// Applies a versioned patch only if the version matches.
  static Map<String, dynamic> apply({
    required Map<String, dynamic> original,
    required List<Map<String, dynamic>> operations,
  }) {
    if (operations.isEmpty || operations[0]['op'] != 'version') {
      throw Exception("Missing version operation at beginning of patch");
    }

    final versionOp = operations[0];
    final from = versionOp['from'];
    final to = versionOp['to'];

    if (original['version'] != from) {
      throw Exception(
          "Version mismatch: expected $from, found ${original['version']}");
    }

    // Remove version op
    final actualOps = operations.sublist(1);
    final updated = _apply(original, actualOps);

    updated['version'] = to;
    return updated;
  }

  /// Internal: Diffs recursively
  static List<Map<String, dynamic>> _diff(
    dynamic oldJson,
    dynamic newJson, [
    String basePath = '',
  ]) {
    final patch = <Map<String, dynamic>>[];

    if (oldJson is Map && newJson is Map) {
      final keys = {...oldJson.keys, ...newJson.keys};
      for (final key in keys) {
        if (key == 'version') continue;

        final path = '$basePath/$key';
        final oldVal = oldJson[key];
        final newVal = newJson[key];

        if (!newJson.containsKey(key)) {
          patch.add({'op': 'remove', 'path': path});
        } else if (!oldJson.containsKey(key)) {
          patch.add({'op': 'add', 'path': path, 'value': newVal});
        } else if (oldVal.runtimeType != newVal.runtimeType) {
          patch.add({'op': 'replace', 'path': path, 'value': newVal});
        } else if (oldVal is Map || oldVal is List) {
          patch.addAll(_diff(oldVal, newVal, path));
        } else if (oldVal != newVal) {
          patch.add({'op': 'replace', 'path': path, 'value': newVal});
        }
      }
    } else if (oldJson is List && newJson is List) {
      final minLen =
          oldJson.length < newJson.length ? oldJson.length : newJson.length;
      for (var i = 0; i < minLen; i++) {
        final path = '$basePath/$i';
        final oldItem = oldJson[i];
        final newItem = newJson[i];

        if (oldItem.runtimeType != newItem.runtimeType) {
          patch.add({'op': 'replace', 'path': path, 'value': newItem});
        } else if (oldItem is Map || oldItem is List) {
          patch.addAll(_diff(oldItem, newItem, path));
        } else if (oldItem != newItem) {
          patch.add({'op': 'replace', 'path': path, 'value': newItem});
        }
      }

      if (newJson.length > oldJson.length) {
        for (var i = oldJson.length; i < newJson.length; i++) {
          patch.add({'op': 'add', 'path': '$basePath/$i', 'value': newJson[i]});
        }
      } else if (oldJson.length > newJson.length) {
        var removeOperations = <Map<String, dynamic>>[];

        for (var i = newJson.length; i < oldJson.length; i++) {
          removeOperations.add({'op': 'remove', 'path': '$basePath/$i'});
        }
        patch.addAll(removeOperations.reversed);
      }
    }

    return patch;
  }

  /// Internal: Applies patch operations
  static Map<String, dynamic> _apply(
    Map<String, dynamic> original,
    List<Map<String, dynamic>> operations,
  ) {
    final updated = jsonDecode(jsonEncode(original));

    dynamic getParent(dynamic root, List<String> segments) {
      dynamic current = root;
      for (int i = 0; i < segments.length - 1; i++) {
        final key = segments[i];
        if (current is Map) {
          current = current[key];
        } else if (current is List) {
          current = current[int.parse(key)];
        }
      }
      return current;
    }

    for (final op in operations) {
      final type = op['op'] as String?;
      final path = op['path'] as String?;
      final value = op['value'];
      final keys = path?.split('/').where((e) => e.isNotEmpty).toList() ?? [];
      final lastKey = keys.last;
      final parent = getParent(updated, keys);

      if (parent is Map) {
        if (type == 'add' || type == 'replace') {
          parent[lastKey] = value;
        } else if (type == 'remove') {
          parent.remove(lastKey);
        }
      } else if (parent is List) {
        final index = int.parse(lastKey);
        if (type == 'add') {
          if (index == parent.length) {
            parent.add(value);
          } else {
            parent.insert(index, value);
          }
        } else if (type == 'replace') {
          parent[index] = value;
        } else if (type == 'remove') {
          parent.removeAt(index);
        }
      }
    }

    return Map<String, dynamic>.from(updated);
  }
}
